### PR TITLE
feat(api): add elasticsearch service

### PR DESCRIPTION
**Pull request by sguiheux**
1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

File: engine/elasticsearch/elasticsearch.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+		WriteTimeout:   10 * time.Minute,
+		MaxHeaderBytes: 1 << 20,
+	}
+
+	//Gracefully shutdown the http server
+	go func() {
+		select {
+		case <-ctx.Done():
+			log.Info("ElasticSearch> Shutdown HTTP Server")
+			server.Shutdown(ctx)
```
Comment (golangci) [inline engine/elasticsearch/elasticsearch.go]: Error return value of `server.Shutdown` is not checked

File: engine/api/services/dao.go [hunk 2]
```diff
@@ -240,3 +255,38 @@ func (r *Repository) FindDeadServices(t time.Duration) ([]sdk.Service, error) {
...
+
+func updateCache(s sdk.Service) {
+	ss, ok := servicesCacheByType[s.Type]
+	indexToUpdate := -1
+	if !ok {
+		ss = make([]sdk.Service, 0, 1)
+	} else {
+		for i, sub := range ss {
+			if sub.Name == s.Name {
+				indexToUpdate = i
```
Comment (bnjjj) [inline engine/api/services/dao.go]: missing break

File: engine/api/services/dao.go [hunk 3]
```diff
@@ -240,3 +255,38 @@ func (r *Repository) FindDeadServices(t time.Duration) ([]sdk.Service, error) {
...
+
+func removeFromCache(s sdk.Service) {
+	ss, ok := servicesCacheByType[s.Type]
+	if !ok || len(ss) == 0 {
+		return
+	}
+	indexToSplit := 0
+	for i, sub := range ss {
+		if sub.Name == s.Name {
+			indexToSplit = i
```
Comment (bnjjj) [inline engine/api/services/dao.go]: missing break

File: engine/api/event/elasticsearch.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+			}
+
+			switch e.EventType {
+			case "sdk.EventPipelineBuild", "sdk.EventJob":
+				continue
+			}
+			e.Payload = nil
+			code, errD := services.DoJSONRequest(esServices, "POST", "/events", e, nil)
+			if code >= 400 || errD != nil {
+				log.Warning("PushInElasticSearch> Unable to send event to elasticsearch [%d]: %v", code, errD)
```
Comment (fsamin) [inline engine/api/event/elasticsearch.go]: log.Error ?

File: engine/api/event/elasticsearch.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+				continue
+			}
+
+			if len(esServices) == 0 {
+				continue
+			}
+
+			var e sdk.Event
+			if err := json.Unmarshal([]byte(msg), &e); err != nil {
+				log.Warning("PushInElasticSearch> Cannot unmarshal event %s: %s", msg, err)
```
Comment (fsamin) [inline engine/api/event/elasticsearch.go]: log.Error
%v for error

File: engine/api/event/elasticsearch.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+		case <-tick.C:
+			msg, err := store.GetMessageFromSubscription(c, pubSub)
+			if err != nil {
+				log.Warning("PushInElasticSearch> Cannot get message %s: %s", msg, err)
+				continue
+			}
+
+			esServices, errS := querier.FindByType(services.TypeElasticsearch)
+			if errS != nil {
+				log.Warning("PushInElasticSearch> Unable to get elasticsearch service")
```
Comment (fsamin) [inline engine/api/event/elasticsearch.go]: log.Error + display the error

File: engine/api/event/elasticsearch.go [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+		select {
+		case <-c.Done():
+			if c.Err() != nil {
+				log.Error("PushInElasticSearch> Exiting: %v", c.Err())
+				return
+			}
+		case <-tick.C:
+			msg, err := store.GetMessageFromSubscription(c, pubSub)
+			if err != nil {
+				log.Warning("PushInElasticSearch> Cannot get message %s: %s", msg, err)
```
Comment (fsamin) [inline engine/api/event/elasticsearch.go]: log.Error ?

File: engine/api/services/dao.go [hunk 8]
```diff
@@ -13,6 +13,12 @@ import (
 	"github.com/ovh/cds/sdk"
 )
 
+var servicesCacheByType map[string][]sdk.Service
+
+func init() {
+	servicesCacheByType = make(map[string][]sdk.Service)
```
Comment (fsamin) [inline engine/api/services/dao.go]: there will be a race condition

File: engine/api/event/elasticsearch.go [hunk 9]
```diff
@@ -29,7 +29,7 @@ func PushInElasticSearch(c context.Context, db gorp.SqlExecutor, store cache.Sto
 
 			esServices, errS := querier.FindByType(services.TypeElasticsearch)
 			if errS != nil {
-				log.Warning("PushInElasticSearch> Unable to get elasticsearch service")
+				log.Error("PushInElasticSearch> Unable to get elasticsearch service")
```
Comment (fsamin) [inline engine/api/event/elasticsearch.go]: add the error in the log

**Pull request merged**